vllm/core/block/interfaces.py

```diff
@@ -91,6 +91,11 @@ def content_hash(self) -> Optional[int]:
 
 class BlockAllocator(ABC):
 
+    num_blocks: int
+    block_size: int
+    block_index_start: int
+    block_index_end: int
```

Comment:
Let's define these as `@property` so implementations not coupled with the interface beyond the required bits

Reply:
I think `@property` is over used in the project. It also has runtime cost, and you need to keep a private `self._num_blocks` . That's not good in my opinion.

Reply:
runtime cost of property is negligible if it helps to create a better/more maintainable API

Reply:
I don't think `@property` is more maintainable here. In fact, I think it adds unnecessary complexity. We need to define `self. _num_blocks` as well as `@property def num_blocks`. If future class needs to define `@property`, they can do it. But we don't need to do it here.

Reply:
We need to prioritize code quality in the block manager because the software design requirements are so complex. Otherwise we will trend towards the V1 block manager and its problems -- untestable, hard to modify thinking about the exponential complexity as prefix caching x cow x swap x lookahead x sliding window are combined.

The concrete things `@property` buys us here:
* We don't allow users to write to these fields (only read). This allows the implementations to trust their fields are not modified, making them easier to reason about. We don't want to have to closely review every open source PR that goes into the block manager to ensure people don't write to these fields.
* We allow future implementations to have different implementations of the same API. this can allow performance optimizations (like how this PR required minimal API changes) or different features. We shouldn't trade this off without good reason.

You are correct that this adds overhead (LOAD_ATTR is replaced by a method call). If this is significantly impactful to ITL then we need to reconsider the approach. Unfortunately, if we cannot maintain good design due to Python performance in a component that requires good design, it means we have to rewrite it in C++. Let's motivate this decision by data before we compromise on design requirements.

You are correct that there is more code (storing private fields in constructor and returning them in property). This is required to allow only reads to those fields and is well worth the tradeoff. I think it's a small amount of code and confined to single files :)

Reply:
I agree with code quality, but we should use the correct tool rather than abusing `@property`. For example, if you want to declear a field is read-only, you can use `from typing import Final`.

Reply:
We can use alternative methods to make a field read-only like type checking. The downsides of using mypy is that it does not have great check coverage and can miss some cases, whereas defining an API ensures users can't go wrong when they're using it.

Reply:
I think we are over-reacting here. Let's make our life easier. I don't want to add unnecessary complexity that we need to write so many lines just to define a field. Who will modify these fields?

Reply:
To summarize, I don't think forcing `@property.getter` is necessary until I see someone violates this rule, which is very rare in my understanding.

Reply:
Unfortunately it is necessary to maintain good design of block manager v2. Don't want to regress on that without good reason. The <1% performance improvement of this PR not a good enough reason.

Reply:
okay, please go ahead then. I don't have bandwidth to follow up with this PR.